frameworks/Clojure/aleph/src/hello/handler.clj

```diff
@@ -1,45 +1,188 @@
 
-(defn handler [req]
+(def html-response
+  {:status 200
+   :headers {"Content-Type" "text/html; charset=utf-8"}})
+
+(def db-spec
+  {:uri "postgresql://tfb-database:5432/hello_world"
+   :user "benchmarkdbuser"
+   :password "benchmarkdbpass"
+   :size 1})
+
+(defmacro thread-local [& body]
+  `(let [tl# (ThreadLocal/withInitial (reify Supplier (get [_] ~@body)))]
+     (reify IDeref (deref [_] (.get tl#)))))
+
+(def pool
+  "PostgreSQL pool of connections (`PgPool`)."
+  (thread-local (async/pool db-spec)))
+
+(defn random
+  "Generate a random number between 1 and 10'000."
+  []
+  (unchecked-inc (.nextInt (ThreadLocalRandom/current) 10000)))
+
+(defn sanitize-queries-param
+  "Sanitizes the `queries` parameter. Clamps the value between 1 and 500.
+  Invalid (string) values become 1."
+  [request]
+  (let [queries (-> request
+                    :query-string
+                    (subs 8))
+        n (try (Integer/parseInt queries)
+               (catch Exception _ 1))] ; default to 1 on parse failure
+    (cond
+      (< n 1) 1
+      (> n 500) 500
+      :else n)))
+
+(def ^Context
+  query-mapper
+  "Map each row into a record."
+  (async/context {:row (async/rs->compiled-record)}))
+
+(defn query-one-random-world
+  "Query a random world on the database.
+  Return a `CompletableFuture`."
+  []
+  (async/query-one query-mapper
+                   @pool
+                   ["SELECT id, randomnumber FROM world WHERE id=$1" (random)]))
+
+(defn update-world
+  "Update a world on the database.
+  Return a `CompletableFuture`."
+  [{:keys [randomNumber id]}]
+  (async/query @pool
+               ["UPDATE world SET randomnumber=$1 WHERE id=$2" randomNumber id]))
+
+(defn run-queries
+  "Run a number of `queries` on the database to fetch a random world.
+  Return a `manifold.deferred`."
+  [queries]
+  (apply d/zip
+         (take queries
+               (repeatedly query-one-random-world))))
+
+(defn query-fortunes
+  "Query the fortunes on database.
+  Return a `CompletableFuture`."
+  []
+  (async/query query-mapper @pool ["SELECT id, message from FORTUNE"]))
+
+(defn get-fortunes
+  "Fetch the full list of Fortunes from the database, sort them by the fortune
+  message text.
+  Return a `CompletableFuture` with the results."
+  []
+  (d/chain (query-fortunes)
+           (fn [fortunes]
+             (sort-by :message
+                      (conj fortunes
+                            {:id 0
+                             :message "Additional fortune added at request time."})))))
+
+(defn update-and-persist
+  "Fetch a number of `queries` random world from the database.
+  Compute a new `randomNumber` for each of them a return a `CompletableFuture`
+  with the updated worlds."
+  [queries]
+  (d/chain' (run-queries queries)
+            (fn [worlds]
+              (let [worlds' (mapv #(assoc % :randomNumber (random)) worlds)]
+                (d/chain' (apply d/zip (mapv update-world worlds'))
```

Comment:
This updates each world independently, even if in parallel, but according to

> Using bulk updates—batches of update statements or individual update statements that affect multiple rows—is acceptable but not required. To be clear: bulk reads are not permissible for selecting/reading the rows, but bulk updates are acceptable for writing the updates.

we could group the statements, which might be faster.